### ISSUE TITLE
URLBear: link_ignore_list syntax unnecessarily complicated

### ISSUE THREAD

**Issue by jayvdb** [coala/coala-bears#2128]
https://github.com/coala/projects/pull/397/commits/20dcfafc327bf7a3b

It shouldnt be necessary to append `/**` ; that should be implied.
It shouldnt be necessary to prepend `https://` ; that should be implied with a default protocol setting, or at most `//` should be required to imply http *and* https to check both work.

**Comment by Makman2** [comment]
Yeah for the list I also would expect that plain domains work :+1:

**Comment by aashraybhandar1** [comment]
How to proceed with this issue?

**Comment by solutusimmensus** [comment]
Let me check I understand the requirements correctly:
* when the link starts with '//', ignore all links matching that link with prepended 'http:' and 'https:':
 _//example.com should ignore links http://example.com and https://example.com_
* when the link ends with '/', ignore all links starting with that link (possibly can be combined with the above):
_http://example.com/ should ignore links like: http://example.com/, http://example.com/test, http://example.com/a/very/long/path_
* otherwise, ignore exactly that link:
_http://example.com ignores only http://example.com_

The path matching function is in coalib, therefore all changes must be done in **URLBear**.

**Comment by Makman2** [comment]
I (and many browsers) do not see any difference in a link ending with a slash or not (browsers usually expand links to contain a final slash). So that's something we shouldn't distinguish resulting in different casrs.

**Comment by Makman2** [comment]
> browsers usually expand links to contain a final slash

Not for all links, but I'm not aware of the exact rules*

**Comment by solutusimmensus** [comment]
Our **Globbing** parser, which is used in **URLBear**, does differentiate between paths ending with slash and without slash:
```
>>> from coalib.parsing.Globbing import fnmatch
>>> fnmatch('http://www.example.com/', 'http://www.example.com')
False
>>>
```

Should I fix Globbing, use different parser, or automatically append slash in matched URL as workaround?

**Comment by Makman2** [comment]
This is totally excepted from globbing, your pattern does not match the first URL because it doesn't contain the trailing slash (or any other wildcard).

> or automatically append slash in matched URL as workaround?

yes that would be ideal imo. However, please lookup somewhere what are the rules here when this should be done (get some inspirations from browsers if possible ;) )

**Comment by solutusimmensus** [comment]
After a quick research, it works like this:
If you input an URL without trailing slash in the browser's address bar and it can't get server response, it tries for the same URL with trailing slash. Browsers can handle URL to directories with and without trailing slashes, but URLs pointing to files **can't** end with slash.

However, the server may treat such endpoints differently. I've written a minimal Flask application to confirm this. Most servers redirect slashless endpoints to slashful ones.

The question is, will there be cases where the user might want to ignore ".../test/", but check for ".../test"?  This might be a twisted, obscure corner case, but it _can_ happen.

If we choose to ignore such cases, the easiest workaround is to check slashless URL twice (second time with appended slash). I'll start working on this solution, but would like to know what you think about it.

**Comment by Makman2** [comment]
> URLs pointing to files

I guess you mean that endpoints having some kind of extension, like `mycloud.com/user/mrT/fileA.txt`.

> The question is, will there be cases where the user might want to ignore ".../test/", but check for ".../test"? This might be a twisted, obscure corner case, but it can happen.

Hmm... I would rather say this case doesn't exist 😅 For this I would rather wait on issue reports from users to see whether such a feature is wanted/needed^^

**Comment by solutusimmensus** [comment]
The issue is done with PR #2220 (awaiting for code review since January)

**Comment by MacBox7** [comment]
I don't understand the need for both `link_ignore_list` and `link_ignore_regex`. Shouldn't a list of `link_ignore_regex` be sufficient?

**Comment by Makman2** [comment]
I guess those just expose two different syntaxes to enumerate links to be ignored. Though this makes the interface unnecessarily bloated and questionable^^
Please file an issue to discuss that. @MacBox7

**Comment by jayvdb** [comment]
The non-regex parameter is intended to be more user friendly. The git history would show it was intentionally created. Unfortunately it wasnt created as intended, hence this current issue.
The regex param allows precision where the new parameter may not allow it.
Reassess after this issue is finished.
